[SPARK-16967] move mesos to module

mgummelt opened the pull request:
## What changes were proposed in this pull request?

Move Mesos code into a mvn module
## How was this patch tested?

unit tests
manually submitting a client mode and cluster mode job
spark/mesos integration test suite

Commit:
move mesos to module

SparkQA commented:
**[Test build #63753 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63753/consoleFull)** for PR 14637 at commit [`57a191b`](https://github.com/apache/spark/commit/57a191bc6b046d68ed1ed9dde8b5c52710fdf214).

jaceklaskowski commented on mesos/pom.xml:
```diff
@@ -0,0 +1,167 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one or more
```
I think the project uses `*` (stars) for continuation lines.

jaceklaskowski commented on mesos/pom.xml:
```diff
@@ -0,0 +1,167 @@
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Isn't the line too long?

mgummelt replied:
I don't think line length rules apply to XML.  See the other poms for examples.

SparkQA commented:
**[Test build #63754 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63754/consoleFull)** for PR 14637 at commit [`465929e`](https://github.com/apache/spark/commit/465929e41a89941dbf5060b68d1c15958df4af91).

SparkQA commented:
**[Test build #63753 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63753/consoleFull)** for PR 14637 at commit [`57a191b`](https://github.com/apache/spark/commit/57a191bc6b046d68ed1ed9dde8b5c52710fdf214).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63753/
Test FAILed.

jaceklaskowski commented on mesos/pom.xml:
```diff
@@ -0,0 +1,167 @@
+    </dependency>
+
+
+    <!-- Explicit listing of transitive deps that are shaded. Otherwise, odd compiler crashes.
```
Why do we need that commented-out section?

mgummelt replied:
This is a WIP.  Please wait to comment.

jaceklaskowski commented on mesos/pom.xml:
```diff
@@ -0,0 +1,167 @@
+      <artifactId>mockito-core</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <!--
```
And that one too?

jaceklaskowski commented on mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterManager.scala:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  override def createSchedulerBackend(sc: SparkContext,
+                                      masterURL: String,
```
I think the convention is to inline 4 spaces only not to align to the first parameter.

jaceklaskowski commented on mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterManager.scala:
```diff
@@ -0,0 +1,61 @@
+                                      masterURL: String,
+                                      scheduler: TaskScheduler): SchedulerBackend = {
+    val mesosUrl = MESOS_REGEX.findFirstMatchIn(masterURL).get.group(1)
+    val coarse = sc.conf.getBoolean("spark.mesos.coarse", defaultValue = true)
```
Oh, I thought Spark supports coarse-grain scheduling only (even for Mesos that had fine-grained one too). Will it stay or is this a temporary thing?

mgummelt replied:
Mesos supports coarse-grained and fine-grained, but fine-grained is deprecated

SparkQA commented:
**[Test build #63754 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63754/consoleFull)** for PR 14637 at commit [`465929e`](https://github.com/apache/spark/commit/465929e41a89941dbf5060b68d1c15958df4af91).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63754/
Test FAILed.

SparkQA commented:
**[Test build #63765 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63765/consoleFull)** for PR 14637 at commit [`4e276ab`](https://github.com/apache/spark/commit/4e276ab81848669ec5c7da6d51cf63a46b1bac87).

SparkQA commented:
**[Test build #63765 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63765/consoleFull)** for PR 14637 at commit [`4e276ab`](https://github.com/apache/spark/commit/4e276ab81848669ec5c7da6d51cf63a46b1bac87).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63765/
Test FAILed.

srowen commented:
Directionally this is looking spot on, thank you. Ping when you want a review.

mgummelt commented on project/SparkBuild.scala:
```diff
@@ -56,9 +56,9 @@ object BuildCommons {
   ).map(ProjectRef(buildLocation, _)) ++ sqlProjects ++ streamingProjects
 
-  val optionallyEnabledProjects@Seq(yarn, java8Tests, sparkGangliaLgpl,
+  val optionallyEnabledProjects@Seq(mesos, yarn, java8Tests, sparkGangliaLgpl,
```
FYI I'm just pattern matching here to get `-Pmesos` to work with `sbt`.  I have no idea what this does.

jaceklaskowski replied:
This is a pattern matching on assignment feature in Scala and the line creates one `optionallyEnabledProjects` value for the entire `Seq` of projects with respective projects being their own values `mesos` etc.

mgummelt commented:
@srowen Ready for review.

SparkQA commented:
**[Test build #64109 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64109/consoleFull)** for PR 14637 at commit [`95e473b`](https://github.com/apache/spark/commit/95e473ba046298d5669dfe63264fc139d600b787).

SparkQA commented:
**[Test build #64109 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64109/consoleFull)** for PR 14637 at commit [`95e473b`](https://github.com/apache/spark/commit/95e473ba046298d5669dfe63264fc139d600b787).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64109/
Test FAILed.

mgummelt commented:
mima seems to be upset about my removal of MESOS_REGEX from `SparkMasterRegex`, but I don't understand why, as it's a private class.  Should I add an entry to MimaExcludes?

srowen commented on mesos/pom.xml:
```diff
@@ -0,0 +1,109 @@
+      </exclusions>
+    </dependency>
+
+    <dependency>
```
I don't know that you need these dependencies -- literally don't know either way by looking. These are definitely needed? I was wondering whether jetty was relevant.

mgummelt replied:
I just now removed mockito, because it was already a dependency of the parent project.

I'm pretty sure all of the shaded deps need to be listed.  I don't know why they're shaded in the first place, but since they are, they must be listed here to avoid classloading errors at runtime.

vanzin replied:
You shouldn't need to touch the shaded dependencies (see how no other module does that). If you're running into problems, you're probably missing something else, so post the error so we can take a look.

mgummelt replied:
YARN does it: https://github.com/apache/spark/blob/master/yarn/pom.xml#L79

which is what I was using as a starting point.  I initially did try to remove them, but got a ClassNotFoundException in WebUI.class when trying to import org.eclipse.jetty.<something>

The parent project seems to rename org.eclipse.jetty to org.spark_project.jetty https://github.com/apache/spark/blob/master/pom.xml#L2273

It's documented that the shaded deps should be put into compile scope in the module: https://github.com/apache/spark/blob/master/pom.xml#L315

Though I still don't understand a) why they're shaded in the first place, and b) how those classes are loaded when spark is built without a module that promotes those dependencies to compile scope.

vanzin replied:
They're shaded because people want to use different versions of guava and jetty.

Spark doesn't need to promote them to compile scope because the relocated guava / jetty classes are packaged with Spark itself.

I see what they problem you're running into is now. It has nothing to do with compile scope, you just need to explicit list the dependencies because the mesos code references them directly. So you actually just need to fix the comment since you're not promoting anything to compile scope by explicitly adding the dependencies (they still inherit the scope from the parent pom, which is "provided", and is actually what you want here).

vanzin replied:
(Note to self: YARN - and this new mesos module - should only need to declare the Guava dependency explicitly, since they don't use Jetty directly.)

mgummelt replied:
Ah, I didn't realize the intransitivity of provided dependencies applied to inheritance as well.

srowen commented:
You can ignore the MiMa error and suppress it, yes.
This is generally looking great, thanks. @vanzin would be a good reviewer too.

I think we'll have to change the Jenkins job config later to set `-Pmesos` -- will see about it.
There are some other miscellaneous places that may need to add the new profile, `dev/`:

- `lint-java`, `scalastyle`
- `mima` add to the profiles it uses for MiMa
- `create-release/release-build.sh` I assume the default tarballs will still build in Mesos

Although it's kind of dead config, you might add the profile to `.travis.yml` too.

vanzin commented:
If this is ready for review could you remove "wip" from the pr title?

vanzin commented:
Also, with the current changes, doesn't it mean the PR builder is not building the mesos code? Shouldn't you modify `dev/run-tests.py` (I think that's the right place)?

The modification to the dependency files also raises another question: how will this be made available to users? Will there be a new package "spark-with-mesos" that needs to be packaged separately? Will users add it with "--packages" instead? There might be changes needed to `dev/create-release/release-build.sh` depending on the answer.

srowen commented:
Good call, I think it's dev/sparktests/modules.py that needs to be modified. it can clone YARN's config.

I assume that the release binaries will have this profile activated as with YARN.

Commit:
add profile to build instructions

mgummelt commented:
@srowen OK, added the profile to:
- .travis.yml
- release-build.sh
- lint-java
- mima
- scalastyle

What about dev/test-dependencies.sh.

Also, I added mesos to PUBLISH_PROFILES in release-build.sh.  Should I also add it to the make_binary_release profiles?

SparkQA commented:
**[Test build #64299 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64299/consoleFull)** for PR 14637 at commit [`3e10acc`](https://github.com/apache/spark/commit/3e10acc733119639bbb4d51c72c5688ddf92bcc6).

srowen commented:
Yes to all three of those.

Commit:
added test module, test-dependencies.sh, and release-build.sh

SparkQA commented:
**[Test build #64301 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64301/consoleFull)** for PR 14637 at commit [`e94f5e7`](https://github.com/apache/spark/commit/e94f5e7c9492ec84b1e2b786f4a2f749eb3250c4).

SparkQA commented:
**[Test build #64301 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64301/consoleFull)** for PR 14637 at commit [`e94f5e7`](https://github.com/apache/spark/commit/e94f5e7c9492ec84b1e2b786f4a2f749eb3250c4).
 * This patch **fails executing the `dev/run-tests` script**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64301/
Test FAILed.

Commit:
fix syntax erro

SparkQA commented:
**[Test build #64302 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64302/consoleFull)** for PR 14637 at commit [`fe7d05e`](https://github.com/apache/spark/commit/fe7d05e94a7cc55c57ba14b041ad3ce07139878e).

SparkQA commented:
**[Test build #64302 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64302/consoleFull)** for PR 14637 at commit [`fe7d05e`](https://github.com/apache/spark/commit/fe7d05e94a7cc55c57ba14b041ad3ce07139878e).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64302/
Test FAILed.

Commit:
replace mesos deps in dependency tests

SparkQA commented:
**[Test build #64303 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64303/consoleFull)** for PR 14637 at commit [`05614b4`](https://github.com/apache/spark/commit/05614b4f8f40c08cb30993a6aa236c6e4cc57276).

SparkQA commented:
**[Test build #64299 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64299/consoleFull)** for PR 14637 at commit [`3e10acc`](https://github.com/apache/spark/commit/3e10acc733119639bbb4d51c72c5688ddf92bcc6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64299/
Test PASSed.

Commit:
add back in mockito

SparkQA commented:
**[Test build #64312 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64312/consoleFull)** for PR 14637 at commit [`cdc5753`](https://github.com/apache/spark/commit/cdc5753d9813f3358625bec1c674f54e0d69835e).

SparkQA commented:
**[Test build #64303 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64303/consoleFull)** for PR 14637 at commit [`05614b4`](https://github.com/apache/spark/commit/05614b4f8f40c08cb30993a6aa236c6e4cc57276).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64303/
Test FAILed.

SparkQA commented:
**[Test build #64312 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64312/consoleFull)** for PR 14637 at commit [`cdc5753`](https://github.com/apache/spark/commit/cdc5753d9813f3358625bec1c674f54e0d69835e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64312/
Test FAILed.

mgummelt commented:
retest this please

SparkQA commented:
**[Test build #64317 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64317/consoleFull)** for PR 14637 at commit [`cdc5753`](https://github.com/apache/spark/commit/cdc5753d9813f3358625bec1c674f54e0d69835e).

SparkQA commented:
**[Test build #64317 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64317/consoleFull)** for PR 14637 at commit [`cdc5753`](https://github.com/apache/spark/commit/cdc5753d9813f3358625bec1c674f54e0d69835e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64317/
Test FAILed.

mgummelt commented:
retest this please

SparkQA commented:
**[Test build #64370 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64370/consoleFull)** for PR 14637 at commit [`cdc5753`](https://github.com/apache/spark/commit/cdc5753d9813f3358625bec1c674f54e0d69835e).

SparkQA commented:
**[Test build #64370 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64370/consoleFull)** for PR 14637 at commit [`cdc5753`](https://github.com/apache/spark/commit/cdc5753d9813f3358625bec1c674f54e0d69835e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64370/
Test PASSed.

mgummelt commented:
@srowen Ready for review

srowen commented:
Nice one, LGTM. I'll leave it open for final comments until tomorrow.

vanzin commented on dev/create-release/release-build.sh:
```diff
@@ -186,12 +186,13 @@ if [[ "$1" == "package" ]]; then
-  make_binary_release "without-hadoop" "-Psparkr -Phadoop-provided -Pyarn" "3038" &
+  FLAGS="-Psparkr -Phadoop-2.3 -Phive -Phive-thriftserver -Pyarn -Pmesos"
+  make_binary_release "hadoop2.3" "$FLAGS" "3033" &
+  make_binary_release "hadoop2.4" "$FLAGS" "3034" &
```
This is wrong now; "FLAGS" enables "-Phadoop-2.3" when here it shuold be "-Phadoop-2.4" (and matching versions in the lines below).

mgummelt replied:
ah, yea.  fixing...

Commit:
fix flags

SparkQA commented:
**[Test build #64430 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64430/consoleFull)** for PR 14637 at commit [`09f3197`](https://github.com/apache/spark/commit/09f3197e7cac9a45315bf5bdaed57c97bcd0e46d).

vanzin commented:
LGTM now.

SparkQA commented:
**[Test build #64430 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64430/consoleFull)** for PR 14637 at commit [`09f3197`](https://github.com/apache/spark/commit/09f3197e7cac9a45315bf5bdaed57c97bcd0e46d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64430/
Test FAILed.

mgummelt commented:
retest this please

SparkQA commented:
**[Test build #64435 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64435/consoleFull)** for PR 14637 at commit [`09f3197`](https://github.com/apache/spark/commit/09f3197e7cac9a45315bf5bdaed57c97bcd0e46d).

SparkQA commented:
**[Test build #64435 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64435/consoleFull)** for PR 14637 at commit [`09f3197`](https://github.com/apache/spark/commit/09f3197e7cac9a45315bf5bdaed57c97bcd0e46d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64435/
Test PASSed.

vanzin commented:
Merging to master.

drcrallen commented:
FYI, we have a build process that packages spark core, now that mesos is is in its own artifact, this broke our build and deploy process, and its not called out in release notes

srowen commented:
The release notes are for end users, and this doesn't impact end users. Developers are expected, more or less, to follow commits and dev@ to keep up with changes like this.
